### PR TITLE
chore: bump and align dependencies; dedupe typescript

### PR DESCRIPTION
**Pull request by legobeat**
This bumps and aligns dependency versions to resolve inefficiencies and fixed upstream issues.

- `madge@6.1.0` currently pulls in multiple different versions of `typescript`. This removes the `3.x` and `4.x` transitive dependencies, aligning on a single `^5.0.4`, which is now also the version used as `devDependency`.
- Remove redundant explicit `detective-*` packages which are already included through `precinct`
- Bump dependencies
  - Some dependency packages have been updated to ESM modules. Updating to latest for these packages will require converting `madge` to ESM as well, which should be done as a future breaking change in madge v7 or later. These have been kept at the latest CommonJS version.
- test: Remove specific typescript configuration for ASP.net

This reduces the size of `node_modules` on a local checkout from 333MB to 149MB.

### CODE REVIEW

**Comment by vbraun** [comment]
This also updates dependencies
```
"dependency-tree": "^10.0.9"
   --> "filing-cabinet": "^4.1.6",
```
which would pull in https://github.com/dependents/node-filing-cabinet/issues/117 which should fix some breakage with index.ts files

**Comment by PabloLION** [comment]
Hi, I'm just back to dev.
Thanks for this great PR. Aligning ts version was what I wanted to do. Great to see someone actually did it. And those `detective-*` are never explicitly imported in the code so removing them are reasonable.
As for the ESM conversion, I remembered that @pahen was against it, right? If not we can do that in v7.

File: test/typescript.js [hunk 1]
```diff
@@ -52,8 +52,7 @@ describe('TypeScript', () => {
 					module: ts.ModuleKind.CommonJS,
 					allowJs: true,
 					configFilePath: undefined
-				},
-				compileOnSave: undefined
```
Comment (PabloLION) [inline test/typescript.js:56]: Could you add an explanation to these lines in the description?

**Review by PabloLION — CHANGES_REQUESTED** [review]

File: test/typescript.js (hunk 1 above)
Comment (PabloLION) [reply test/typescript.js:56]: I supposed that, but when I check [TS5.0 release notes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html) and current [tsconfig definition](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html), ([permalink](https://github.com/microsoft/TypeScript-Website/blob/9867af1a2a53c669c7a4d0be6dd1fa38a665bc47/packages/documentation/copy/en/project-config/tsconfig.json.md)), there were no occurrence of `compileOnSave`.
In contrary, it seems like a Visual Studio feature [see this link](https://www.typescriptlang.org/docs/handbook/asp-net-core.html).
So maybe it's not needed include it in our test?

Comment (PabloLION) [reply test/typescript.js:56]: I thought it the opposite way. Nice catch!

**Review by PabloLION — APPROVED** [review]

**Comment by PabloLION** [comment]
Hi, @legobeat ! thanks for the PR and I think it's almost ready to get merged! Only thing is that I find the base of your branch is 499 commits ago. Could you rebase it to current master? I tried to do it but it seems you didn't select "Allow edits from maintainers" so my hands are tied. 
Thanks for the PR again!

**Comment by PabloLION** [comment]
Sorry I just logged off. This will be the first priority on my next session. Thanks again!

**Comment by PabloLION** [comment]
I found TS>5.2.0 gets a warning (no error) with @typescript-eslint/typescript-estree So I'm limiting it to 5.2.0. I think it's enough for our current stage. And since it's already in the dev dependency, we no longer need it as a peer dep nor in the peer dep meta. So I removed those fields.
How does this look to you @legobeat ? If you are ok with these changes, I'll try to merge it now.

**Comment by PabloLION** [comment]
Sorry that I was sloppy on [9db3ebd](https://github.com/pahen/madge/pull/379/commits/9db3ebd0abf53e233e9034d994545cdc32b89b72)

Your statements are all correct, I'm just trying to relate them to `[726d2c2](https://github.com/pahen/madge/commit/726d2c213949c33b6636c52c97bc82c540fb1127)` 

> * `@typescript-eslint/typescript-estree` is only used as a `devDependency` for linting.
>   * While it seems reasonable to restrict the version used as `devDependency`, this shouldn't affect what versions can be used for runtime.
We should restrict typescript version as devDependency

> * `devDependencies` is not recognized for users / dependents of the package. They are only relevant for devs, ie when running `${PACKAGEMANAGER} install` in the repo.
> * Did anything other related to this change since [Move typescript to peer dependencies #350](https://github.com/pahen/madge/pull/350) / [constrain and bump typescript versions #376](https://github.com/pahen/madge/pull/376)? Otherwise seems like the motivations for keeping it as optional `peerDependency` would still be as valid?
Should keep TS as peerDependency for non-dev users.

Finally, thanks for the correction and the work.

**Pull request merged**